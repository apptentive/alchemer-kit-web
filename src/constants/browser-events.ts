/**
 * Constants
 */
export default {
  APPTENTIVE_APP_STORE_RATING_BLOCKED: 'apptentive:app-store-rating:blocked',
  APPTENTIVE_APP_STORE_RATING_ERROR: 'apptentive:app-store-rating:error',
  APPTENTIVE_APP_STORE_RATING_OPEN_APP_STORE_URL: 'apptentive:app-store-rating:open-app-store-url',
  APPTENTIVE_ERROR: 'apptentive:error',
  APPTENTIVE_INIT: 'apptentive:init',
  APPTENTIVE_READY: 'apptentive:ready',
  APPTENTIVE_LOVE_DIALOG_CANCEL: 'apptentive:love-dialog:cancel',
  APPTENTIVE_LOVE_DIALOG_CLOSE: 'apptentive:love-dialog:close',
  APPTENTIVE_LOVE_DIALOG_DISMISS: 'apptentive:love-dialog:dismiss',
  APPTENTIVE_LOVE_DIALOG_ERROR: 'apptentive:love-dialog:error',
  APPTENTIVE_LOVE_DIALOG_LAUNCH: 'apptentive:love-dialog:launch',
  APPTENTIVE_LOVE_DIALOG_NO: 'apptentive:love-dialog:no',
  APPTENTIVE_LOVE_DIALOG_YES: 'apptentive:love-dialog:yes',
  APPTENTIVE_MESSAGE_CENTER_CLOSE: 'apptentive:message-center:close',
  APPTENTIVE_MESSAGE_CENTER_DISMISS: 'apptentive:message-center:dismiss',
  APPTENTIVE_MESSAGE_CENTER_ERROR: 'apptentive:message-center:error',
  APPTENTIVE_MESSAGE_CENTER_LAUNCH: 'apptentive:message-center:launch',
  APPTENTIVE_MESSAGE_CENTER_SEND: 'apptentive:message-center:send',
  APPTENTIVE_MESSAGES_CREATED: 'apptentive:messages:created',
  APPTENTIVE_MESSAGES_FETCHED: 'apptentive:messages:fetched',
  APPTENTIVE_NAVIGATE_TO_LINK_BLOCKED: 'apptentive:navigate-to-link:blocked',
  APPTENTIVE_NAVIGATE_TO_LINK_ERROR: 'apptentive:navigate-to-link:error',
  APPTENTIVE_NAVIGATE_TO_LINK_NAVIGATE: 'apptentive:navigate-to-link:navigate',
  APPTENTIVE_NAVIGATE_TO_LINK_NAVIGATED: 'apptentive:navigate-to-link:navigated',
  APPTENTIVE_NOTE_CANCEL: 'apptentive:note:cancel',
  APPTENTIVE_NOTE_CLICK: 'apptentive:note:click',
  APPTENTIVE_NOTE_CLOSE: 'apptentive:note:close',
  APPTENTIVE_NOTE_DISMISS: 'apptentive:note:dismiss',
  APPTENTIVE_NOTE_ERROR: 'apptentive:note:error',
  APPTENTIVE_NOTE_LAUNCH: 'apptentive:note:launch',
  APPTENTIVE_OVERLAY_CLICK: 'apptentive:overlay:click',
  APPTENTIVE_RENDER_INTERACTION: 'apptentive:render:interaction',
  APPTENTIVE_SURVEY_DISMISS: 'apptentive:survey:dismiss',
  APPTENTIVE_SURVEY_ERROR: 'apptentive:survey:error',
  APPTENTIVE_SURVEY_LAUNCH: 'apptentive:survey:launch',
  APPTENTIVE_SURVEY_SUBMIT: 'apptentive:survey:submit',
  APPTENTIVE_SURVEY_NEXT_QUESTION_SET: 'apptentive:survey:next_question_set',
};
