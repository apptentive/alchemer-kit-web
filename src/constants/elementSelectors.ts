export enum surveySelectors {
  title = '.apptentive-appbar__title',
  answerChoice = 'answer-choice',
  answerChoices = 'answer-choices',
  errorMessage = 'apptentive-errors',
  invalidReason = 'invalid-reason',
  thankYou = 'apptentive-survey-thank-you',
  questionsForm = 'apptentive-survey-questions',
  questionsWrapper = 'apptentive-survey__question-wrapper',
  actionsWrapper = 'apptentive-survey__actions-wrapper',
}

export enum messageCenterSelectors {
  content = '.apptentive-message-center__content',
  feedback = '.apptentive-message-center-feedback',
  thankYou = 'apptentive-message-center-thank-you',
  profile = '.apptentive-message-center-profile',
  profileNameInput = '.apptentive-message-center-profile__name .apptentive-textinput',
  profileEmailInput = '.apptentive-message-center-profile__email .apptentive-textinput',
  submitButton = '.apptentive-message-center__actions .apptentive-button--primary',
}

export enum interactionSelectors {
  confirmBlock = 'apptentive-interaction__confirm-wrapper',
  blocked = 'apptentive-interaction__blocked',
}
