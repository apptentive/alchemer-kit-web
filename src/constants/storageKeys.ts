export enum sessionStorageKeys {
  manifestCache = 'apptentiveCachedManifest',
  manifestExpiration = 'apptentiveCacheExpiration',
  manifestLocale = 'apptentiveLocaleOverride',
}

export enum localStorageKeys {
  appId = 'app_id',
  captureDisabled = 'captureDisabled',
  codePoint = 'code_point',
  conversation = 'conversation',
  device = 'device',
  interactionCounts = 'interaction_counts',
  person = 'person',
  random = 'random',
}
