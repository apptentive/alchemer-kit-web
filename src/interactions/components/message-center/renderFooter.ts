/**
 * Renders a given Message Center footer.
 *
 * @returns {HTMLElement} - The footer DOM Node.
 */
const renderFooter = () => {
  const footer = document.createElement('footer');
  footer.classList.add('apptentive-footer');

  return footer;
};

export default renderFooter;
