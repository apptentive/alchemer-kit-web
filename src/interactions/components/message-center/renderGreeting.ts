/**
 * Renders the Message Center greeting (icon, title, body, info link).
 *
 * @param {string} titleText - The text to use as the section title
 * @param {string} bodyText - The text to use for the section content
 * @returns {HTMLDivElement} - The greeting DOM Node
 */
const renderGreeting = (titleText: string, bodyText: string) => {
  // NOTE: Currently unused.
  // const icon = document.createElement('img');
  // icon.className = 'apptentive-message-center-greeting-icon';
  // icon.src = this.sdk.i18n.translate(`${this.interaction.id}.configuration.greeting.image_url`);
  // icon.alt = '';
  // icon.width = 80;
  // icon.height = 80;
  // greeting.appendChild(icon);

  const title = document.createElement('h2');
  title.classList.add('apptentive-message-center-greeting__title');
  title.classList.add('apptentive-typography__header');
  title.textContent = titleText;

  const body = document.createElement('p');
  body.classList.add('apptentive-typography__body');
  body.textContent = bodyText;

  const greeting = document.createElement('div');
  greeting.classList.add('apptentive-message-center-greeting');
  greeting.append(title);
  greeting.append(body);

  return greeting;
};

export default renderGreeting;
