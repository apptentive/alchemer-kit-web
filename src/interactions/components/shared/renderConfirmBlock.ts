import { interactionSelectors } from '../../../constants/elementSelectors';

type BtnListener = (e: MouseEvent) => void;

const overlayListener = (e: MouseEvent) => {
  e.preventDefault();
  e.stopImmediatePropagation();
};

/**
 * Renders the confirmation exit block under interaction header.
 * @param {HTMLElement} interactionContainer - The root DOM node of interaction
 * @param {HTMLElement} container - The DOM node that will be disabled under overlay block
 * @param {string} title - The title to render in the confirmation exit block
 * @param {Function} onSbmt - The method to invoke when the OK button is clicked
 * @param {Function} onCancel - The method to invoke when the CANCEL button is clicked
 * @param {HTMLElement[]} wrapperList - The array of DOM nodes to which className with 0.5 opacity will be aded
 */
function renderConfirmBlock(
  interactionContainer: HTMLElement,
  container: HTMLElement,
  title: string,
  onSbmt: BtnListener,
  onCancel: BtnListener,
  wrapperList?: (HTMLElement | null)[]
) {
  const minimizeBtn = document.querySelector('.apptentive-appbar__action--minimize--active');
  if (minimizeBtn) {
    interactionContainer.className = 'fixed corner';
    minimizeBtn.classList.toggle('apptentive-appbar__action--minimize--active');
  }

  const overlay = document.createElement('div');
  overlay.classList.add('apptentive-interaction__overlay');

  overlay.addEventListener('click', overlayListener);
  container.append(overlay);

  wrapperList?.forEach((wrapper: HTMLElement | null) => wrapper?.classList.add(interactionSelectors.blocked));

  const confirmWrapper = document.querySelector<HTMLElement>(`.${interactionSelectors.confirmBlock}`);
  confirmWrapper?.classList.replace('closed', 'opened');

  const titleNode = document.createElement('span');
  titleNode.textContent = title;

  const okBtn = document.createElement('button');
  const cancelBtn = document.createElement('button');

  okBtn.classList.add('apptentive-interaction__confirm-wrapper__ok');
  okBtn.textContent = 'OK';
  okBtn.addEventListener('click', onSbmt);

  cancelBtn.classList.add('apptentive-interaction__confirm-wrapper__cancel');
  cancelBtn.textContent = 'CANCEL';

  const cleanUp = () => {
    overlay.remove();
    titleNode.remove();
    okBtn.remove();
    cancelBtn.remove();
  };

  cancelBtn.addEventListener('click', (e) => {
    onCancel(e);
    cleanUp();
  });

  confirmWrapper?.append(titleNode);
  confirmWrapper?.append(okBtn);
  confirmWrapper?.append(cancelBtn);
}

export default renderConfirmBlock;
