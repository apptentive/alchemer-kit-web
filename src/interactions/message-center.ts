/* exported ApptentiveMessageCenter */
import browserEvent from '../browser-event';
import browserEvents from '../constants/browser-events';
import internalEvents from '../constants/events';
import ApptentiveInteraction from './interaction-base';
import type ApptentiveBase from '../base';
import { IMessageCenterConfiguration } from '../interfaces/interactions/IMessageCenterConfiguration';
import { IInteraction } from '../interfaces/manifest/IInteraction';
import { IPerson } from '../interfaces/data/IPerson';
import { IBaseInteractionOptions } from '../interfaces/interactions/IBaseInteractionOptions';
import { IErrorResponse } from '../interfaces/api/IErrorResponse';
import renderFooter from './components/message-center/renderFooter';
import renderThankYou from './components/shared/renderThankYou';
import { interactionSelectors, messageCenterSelectors } from '../constants/elementSelectors';
import renderIntroduction from './components/message-center/renderIntroduction';
import renderGreeting from './components/message-center/renderGreeting';
import renderTextArea from './components/message-center/renderTextArea';
import renderSubmitButton from './components/message-center/renderSubmitButton';
import renderProfile from './components/message-center/renderProfile';
import renderAppbar from './components/shared/renderAppbar';
import renderConfirmBlock from './components/shared/renderConfirmBlock';

export default class ApptentiveMessageCenter extends ApptentiveInteraction<IMessageCenterConfiguration> {
  private missingPersonInfo!: boolean;
  private wantProfile!: boolean;
  private isConfirmBlock: boolean;

  public container!: HTMLElement;
  /**
   * @param {object} interaction - The Message Center object.
   * @param {object} interaction.configuration - The configuration used to attempt to show the Message Center.
   * @param {object} interaction.configuration.automated_message - Automatic Message configuration.
   * @param {string} interaction.configuration.automated_message.body - Automatic Message body text.
   * @param {object} interaction.configuration.greeting - The configuration used to customize the greeting section.
   * @param {string} interaction.configuration.greeting.image_url - URL to the image to show, usually an avatar. (Unused)
   * @param {string} interaction.configuration.greeting.title - The text title to render for the greeting section.
   * @param {string} interaction.configuration.greeting.body - The text body to render for the greeting section
   * @param {string} interaction.configuration.position - The placement CSS classname to position the dialog.
   * @param {object} interaction.configuration.profile - The configuration specific to profile (name & email) collection.
   * @param {boolean} interaction.configuration.profile.require - Requires the profile collection.
   * @param {boolean} interaction.configuration.profile.request - Requests the profile collection, but is not required.
   * @param {object} interaction.configuration.profile.edit - Default labels for editing profile information.
   * @param {string} interaction.configuration.profile.edit.title - Default label for editing profile information section title.
   * @param {string} interaction.configuration.profile.edit.name_hint - Default label for editing profile information name hint text.
   * @param {string} interaction.configuration.profile.edit.email_hint - Default label for editing profile information email hint text.
   * @param {string} interaction.configuration.profile.edit.skip_button - Default label for editing profile information skip button.
   * @param {string} interaction.configuration.profile.edit.save_button - Default label for editing profile information save button.
   * @param {object} interaction.configuration.profile.initial - Default labels for initial collection of profile information.
   * @param {string} interaction.configuration.profile.initial.title - Default label for initial collection of profile information section title.
   * @param {string} interaction.configuration.profile.initial.name_hint - Default label for initial collection of profile information name hint text.
   * @param {string} interaction.configuration.profile.initial.email_hint - Default label for initial collection of profile information email hint text.
   * @param {string} interaction.configuration.profile.initial.skip_button - Default label for initial collection of profile information skip button.
   * @param {string} interaction.configuration.profile.initial.save_button - Default label for initial collection of profile information save button.
   * @param {object} interaction.configuration.status - The configuration specific to status updates.
   * @param {string} interaction.configuration.status.body - The text status to render.
   * @param {string} interaction.configuration.title - The text title to render.
   * @param {number} interaction.version - The version of the interaction, 1 being legacy and 2 the most recent.
   * @param {object} sdk - The Apptentive WebSDK object.
   * @param {object} [options] - Your configuration options.
   */
  constructor(
    interaction: IInteraction<IMessageCenterConfiguration>,
    sdk: ApptentiveBase,
    options: IBaseInteractionOptions
  ) {
    super(interaction, sdk, options);

    this.renderContent = this.renderContent.bind(this);
    this.submit = this.submit.bind(this);
    this.submitted = this.submitted.bind(this);
    this.validateForm = this.validateForm.bind(this);
    this.errored = this.errored.bind(this);
    this.saveProfile = this.saveProfile.bind(this);
    this.render = this.render.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.close = this.close.bind(this);

    this.isConfirmBlock = false;

    if (!interaction || !interaction.configuration) {
      const error = 'Invalid Message Center provided.';
      browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_ERROR, { error });
      this.sdk.console('error', error);
      return;
    }

    if (interaction.version && interaction.version < 2) {
      const error = `Incompatible Message Center version provided: ${interaction.version}`;
      browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_ERROR, { error });
      this.sdk.console('error', error);
      return;
    }

    this.missingPersonInfo = !sdk.logicEngine?.person?.name || !sdk.logicEngine?.person?.email;
    this.wantProfile = interaction.configuration.profile?.require || interaction.configuration.profile?.request;
  }

  /**
   * Renders the Message Center container and content (title, yes, no).
   * @returns {HTMLElement} - The entire container DOM Node.
   */
  renderContent() {
    const content = document.createElement('div');
    content.classList.add('apptentive-message-center__content');

    content.append(
      renderIntroduction(this.sdk.i18n.translate(`${this.interaction.id}.configuration.automated_message.body`))
    );

    content.append(
      renderGreeting(
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.greeting.title`) || 'Hello!',
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.greeting.body`)
      )
    );

    // NOTE: Messages can be shown, but given the restrictions of the web they are unused currently.
    // content.appendChild(this.renderMessages());

    content.append(renderTextArea(this.sdk.i18n.translate(`${this.interaction.id}.configuration.composer.hint_text`)));

    // Only render the profile section if the config requests it and we don't have that information already
    if (this.wantProfile && this.missingPersonInfo) {
      content.append(
        renderProfile({
          isRequired: this.interaction.configuration.profile.require,
          title: this.sdk.i18n.translate(`${this.interaction.id}.configuration.profile.initial.title`),
          name: {
            placeholder: this.sdk.i18n.translate(`${this.interaction.id}.configuration.profile.initial.name_hint`),
            value: this.sdk && this.sdk.logicEngine.person ? this.sdk.logicEngine.person.name || '' : '',
          },
          email: {
            placeholder: this.sdk.i18n.translate(`${this.interaction.id}.configuration.profile.initial.email_hint`),
            value: this.sdk && this.sdk.logicEngine.person ? this.sdk.logicEngine.person.email || '' : '',
          },
        })
      );
    }

    content.append(
      renderSubmitButton(
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.composer.send_button`),
        this.submit
      )
    );

    return content;
  }

  /**
   * Validates the form is ready to be submitted.
   * @returns {boolean} - Whether the form is valid.
   */
  validateForm() {
    let isValid = true;

    // If email is required and there isn't an identified person, don't let the submit through if it isn't filled
    if (this.interaction.configuration.profile?.require && this.missingPersonInfo) {
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
      const emailElement = this.container.querySelector<HTMLInputElement>(messageCenterSelectors.profileEmailInput);
      const emailValue = emailElement?.value.trim();

      if ((!emailValue || !emailRegex.test(emailValue)) && emailElement) {
        isValid = false;
        emailElement.classList.add('apptentive-textinput--error');
      }
    }

    const messageElement = this.container.querySelector<HTMLTextAreaElement>(
      `${messageCenterSelectors.feedback} textarea`
    );

    if (messageElement && !messageElement.value) {
      isValid = false;
      messageElement.classList.add('apptentive-textinput--error');
    }

    return isValid;
  }

  /**
   * Submits the message.
   */
  submit() {
    const isFormValid = this.validateForm();

    if (!isFormValid) {
      return;
    }

    if (this.wantProfile && this.missingPersonInfo) {
      this.saveProfile();
    }

    // The validateForm method makes sure this element exists and it has a value
    const message = this.container
      .querySelector<HTMLTextAreaElement>(`${messageCenterSelectors.feedback} textarea`)!
      .value.trim();

    this.container.querySelector<HTMLButtonElement>(messageCenterSelectors.submitButton)!.disabled = true;
    this.sdk.createMessage({
      body: message,
      success: this.submitted,
      failure: this.errored,
    });
  }

  /**
   * Handle a successful submission by showing a thank you message or dismissing the view.
   * @private
   */
  submitted() {
    this.sdk.engage(internalEvents.APPTENTIVE_MESSAGE_CENTER_SEND, {
      id: this.interaction.id,
      interaction_id: this.interaction.id,
    });

    if (this.interaction.configuration.status && this.interaction.configuration.status.body) {
      this.container.querySelector<HTMLElement>(messageCenterSelectors.content)!.style.display = 'none';
      this.container.querySelector<HTMLElement>(messageCenterSelectors.thankYou)!.style.display = '';
    } else {
      this.handleClose();
    }

    browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_SEND, { interaction_id: this.interaction.id });
  }

  /**
   * Handle an unsuccessful submission by showing the error string.
   * @param {object} data - The failed response object.
   * @param {string} data.responseText - The failed response text.
   * @private
   */
  errored(data: IErrorResponse) {
    this.container.querySelector<HTMLElement>(messageCenterSelectors.content)!.style.display = '';
    this.container.querySelector<HTMLElement>(messageCenterSelectors.thankYou)!.style.display = 'none';

    const errors = this.container.querySelector<HTMLElement>('apptentive-errors');
    let errorText = 'Please double check your message and try again.';

    if (data && data.responseText) {
      try {
        const response = JSON.parse(data.responseText);

        if (response && response.error) {
          errorText = response.error;
        }
      } catch (error) {
        this.sdk.console('error', `Error Submitting: ${error}`);
      }
    }

    // Clean up error text.
    errorText = errorText.trim();

    if (errors) {
      errors.textContent = errorText;
      errors.style.whiteSpace = 'inherit';
      errors.style.display = '';
      errors.scrollIntoView();
    }

    const submitButton = this.container.querySelector<HTMLButtonElement>(messageCenterSelectors.submitButton);

    if (submitButton) {
      submitButton.disabled = false;
    }

    browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_ERROR, { error: errorText });
  }

  /**
   * Updates the name and email for the person object on the server.
   */
  saveProfile() {
    const person = {} as IPerson;
    const name = this.container.querySelector<HTMLInputElement>(messageCenterSelectors.profileNameInput)?.value.trim();
    const email = this.container
      .querySelector<HTMLInputElement>(messageCenterSelectors.profileEmailInput)
      ?.value.trim();

    if (name) {
      person.name = name;
    }

    if (email) {
      person.email = email;
    }

    if (name || email) {
      this.sdk.updatePerson(person);
    }
  }

  /**
   * Renders a given Message Center.
   * @private
   */
  render() {
    if (!this.interaction || (this.interaction && (!this.interaction.id || !this.interaction.configuration))) {
      const error = 'You cannot render a Message Center without providing a valid Message Center configuration object.';
      browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_ERROR, { error });
      this.sdk.console('error', `${error} You provided:`, this.interaction);
      return;
    }

    if (document.querySelectorAll(`#apptentive-message-center-${this.interaction.id}`).length > 0) {
      this.sdk.console('warn', `Message Center ${this.interaction.id} is already rendered.`);
      return;
    }

    this.container = document.createElement('apptentive-message-center');
    this.container.className = `fixed ${this.interaction.configuration.position || 'corner'}`;
    this.container.id = `apptentive-message-center-${this.interaction.id}`;

    this.container.append(
      renderAppbar(
        this.container,
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.title`) || 'Message Center',
        this.interaction.configuration.position,
        this.handleClose
      )
    );

    const confirmWrapper = document.createElement('div');
    confirmWrapper.classList.add(interactionSelectors.confirmBlock);
    confirmWrapper.classList.add('closed');
    this.container.append(confirmWrapper);

    this.container.append(this.renderContent());
    this.container.append(
      renderThankYou(
        this.handleClose,
        messageCenterSelectors.thankYou,
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.status.body`) || 'Thank you!',
        this.sdk.i18n.translate(`${this.interaction.id}.configuration.composer.close_text`) || 'Close',
        false
      )
    );
    this.container.append(renderFooter());

    this.domNode.append(this.container);
    this.container.focus();

    browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_LAUNCH);
    this.sdk.engage(internalEvents.APPTENTIVE_MESSAGE_CENTER_LAUNCH, {
      id: this.interaction.id,
      interaction_id: this.interaction.id,
    });
  }

  /**
   * Close the Message Center or render confirmation bar and engage required event.
   */
  handleClose() {
    if (!this.container) {
      return;
    }

    const onSbmt = (e?: MouseEvent) => {
      e?.preventDefault();
      if (this.sdk) {
        browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_DISMISS);
      }

      this.close();
    };

    if (this.isConfirmBlock) {
      onSbmt();
      return;
    }

    const isTYScreen = this.container.querySelector<HTMLElement>(messageCenterSelectors.thankYou)!.style.display === '';
    const messageElement = this.container.querySelector<HTMLTextAreaElement>(
      `${messageCenterSelectors.feedback} textarea`
    );

    const msgValue = messageElement?.value.trim();

    if (msgValue && !isTYScreen) {
      this.isConfirmBlock = true;

      const contentWrapper = document.querySelector<HTMLElement>(messageCenterSelectors.content) || null;
      const confirmWrapper = document.querySelector<HTMLElement>(
        `.${interactionSelectors.confirmBlock}`
      ) as HTMLElement;

      const onCancel = (e: MouseEvent) => {
        e.preventDefault();
        this.isConfirmBlock = false;
        contentWrapper?.classList.remove(interactionSelectors.blocked);
        confirmWrapper?.classList.replace('opened', 'closed');
      };

      renderConfirmBlock(
        this.container,
        contentWrapper as HTMLElement,
        'Progress will be lost. Would you like to exit?',
        onSbmt,
        onCancel,
        [contentWrapper]
      );

      return;
    }

    if (this.sdk) {
      browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_DISMISS);
    }

    this.close();
  }

  /**
   * Dismiss a rendered Message Center from view.
   */
  close() {
    if (this.container) {
      if (this.sdk) {
        browserEvent(browserEvents.APPTENTIVE_MESSAGE_CENTER_CLOSE);
        this.sdk.engage(internalEvents.APPTENTIVE_MESSAGE_CENTER_CLOSE, {
          id: this.interaction.id,
          interaction_id: this.interaction.id,
        });
      }
      this.container.remove();
    }
  }

  /**
   * Construct and render a Message Center helper method.
   * @param {object} interaction - The Message Center object.
   * @param {object} interaction.configuration - The configuration used to attempt to show the Message Center.
   * @param {object} interaction.configuration.automated_message - Automatic Message configuration.
   * @param {string} interaction.configuration.automated_message.body - Automatic Message body text.
   * @param {object} interaction.configuration.greeting - The configuration used to customize the greeting section.
   * @param {string} interaction.configuration.greeting.image_url - URL to the image to show, usually an avatar. (Unused)
   * @param {string} interaction.configuration.greeting.title - The text title to render for the greeting section.
   * @param {string} interaction.configuration.greeting.body - The text body to render for the greeting section
   * @param {string} interaction.configuration.position - The placement CSS classname to position the dialog.
   * @param {object} interaction.configuration.profile - The configuration specific to profile (name & email) collection.
   * @param {boolean} interaction.configuration.profile.require - Requires the profile collection.
   * @param {boolean} interaction.configuration.profile.request - Requests the profile collection, but is not required.
   * @param {object} interaction.configuration.profile.edit - Default labels for editing profile information.
   * @param {string} interaction.configuration.profile.edit.title - Default label for editing profile information section title.
   * @param {string} interaction.configuration.profile.edit.name_hint - Default label for editing profile information name hint text.
   * @param {string} interaction.configuration.profile.edit.email_hint - Default label for editing profile information email hint text.
   * @param {string} interaction.configuration.profile.edit.skip_button - Default label for editing profile information skip button.
   * @param {string} interaction.configuration.profile.edit.save_button - Default label for editing profile information save button.
   * @param {object} interaction.configuration.profile.initial - Default labels for initial collection of profile information.
   * @param {string} interaction.configuration.profile.initial.title - Default label for initial collection of profile information section title.
   * @param {string} interaction.configuration.profile.initial.name_hint - Default label for initial collection of profile information name hint text.
   * @param {string} interaction.configuration.profile.initial.email_hint - Default label for initial collection of profile information email hint text.
   * @param {string} interaction.configuration.profile.initial.skip_button - Default label for initial collection of profile information skip button.
   * @param {string} interaction.configuration.profile.initial.save_button - Default label for initial collection of profile information save button.
   * @param {object} interaction.configuration.status - The configuration specific to status updates.
   * @param {string} interaction.configuration.status.body - The text status to render.
   * @param {string} interaction.configuration.title - The text title to render.
   * @param {number} interaction.version - The version of the interaction, 1 being legacy and 2 the most recent.
   * @param {object} sdk - The Apptentive WebSDK object.
   * @param {object} [options] - Your configuration options.
   * @static
   */
  static display(
    interaction: IInteraction<IMessageCenterConfiguration>,
    sdk: ApptentiveBase,
    options: IBaseInteractionOptions
  ) {
    try {
      const messageCenter = new ApptentiveMessageCenter(interaction, sdk, options);
      messageCenter.render();
    } catch (_) {}
  }
}
