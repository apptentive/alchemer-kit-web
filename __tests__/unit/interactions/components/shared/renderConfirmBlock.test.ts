import renderConfirmBlock from '../../../../../src/interactions/components/shared/renderConfirmBlock';
import ApptentiveBaseBuilder from '../../../../mocks/builders/ApptentiveBaseBuilder';
import { fullMessageCenterConfig } from '../../../../mocks/data/message-center';
import ApptentiveMessageCenter from '../../../../../src/interactions/message-center';
import { _interactionOptions } from '../../../../mocks/data/shared-constants';
import { messageCenterSelectors } from '../../../../../src/constants/elementSelectors';

const _title = 'Progress will be lost. Would you like to exit?';
const _onSbmt = jest.fn();
const _onCancel = jest.fn();

const buildInteraction = () => {
  const { sdk } = new ApptentiveBaseBuilder()
    .useCachedManifest()
    .useOptions({ interactions: [fullMessageCenterConfig] })
    .build();

  const messageCenter = new ApptentiveMessageCenter(fullMessageCenterConfig, sdk, _interactionOptions);
  messageCenter.render();

  return {
    interactionContainer: messageCenter.container,
    container: document.querySelector<HTMLElement>(messageCenterSelectors.content) as HTMLElement,
  };
};

describe('renderConfirmBlock', () => {
  afterEach(() => {
    document.body.innerHTML = '';
    _onCancel.mockClear();
    _onSbmt.mockClear();
  });

  test('properly renders confirm block', () => {
    const { interactionContainer, container } = buildInteraction();
    renderConfirmBlock(interactionContainer, container, _title, _onSbmt, _onCancel);

    expect(interactionContainer).toMatchSnapshot();
  });

  test('properly attaches submit function', () => {
    const { interactionContainer, container } = buildInteraction();
    renderConfirmBlock(interactionContainer, container, _title, _onSbmt, _onCancel);

    const confirmWrapper = document.querySelector('.apptentive-interaction__confirm-wrapper') as HTMLElement;
    const okBtn = confirmWrapper.childNodes[1] as HTMLElement;
    okBtn.click();

    expect(_onSbmt).toHaveBeenCalledTimes(1);
  });

  test('properly attaches cancel function', () => {
    const { interactionContainer, container } = buildInteraction();
    renderConfirmBlock(interactionContainer, container, _title, _onSbmt, _onCancel);

    const confirmWrapper = document.querySelector('.apptentive-interaction__confirm-wrapper') as HTMLElement;
    const cancelBtn = confirmWrapper.childNodes[2] as HTMLElement;
    cancelBtn.click();

    expect(_onCancel).toHaveBeenCalledTimes(1);
  });

  test('open interaction with confirm exit block on cross btn click when was minimized', () => {
    const { interactionContainer, container } = buildInteraction();
    expect(interactionContainer.className).toBe('fixed corner');

    const minimizeBtn = document.querySelector('.apptentive-appbar__action--minimize') as HTMLElement;
    minimizeBtn.click();

    expect(interactionContainer.className).toBe('fixed corner corner--minimized');

    renderConfirmBlock(interactionContainer, container, _title, _onSbmt, _onCancel);
    expect(interactionContainer.className).toBe('fixed corner');
  });
});
