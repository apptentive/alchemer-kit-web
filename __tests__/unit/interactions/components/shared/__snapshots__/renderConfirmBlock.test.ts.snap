// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderConfirmBlock properly renders confirm block 1`] = `
<apptentive-message-center
  class="fixed corner"
  id="apptentive-message-center-55bae8ef35878bc3c6000582"
>
  <div
    class="apptentive-appbar"
  >
    <div
      class="apptentive-appbar__icon"
    />
    <h1
      class="apptentive-appbar__title"
      tabindex="0"
    >
      Message Center
    </h1>
    <div
      class="apptentive-appbar__actions"
    >
      <div
        aria-label="minimize"
        class="apptentive-appbar__action apptentive-appbar__action--minimize"
        id="message-center__action--minimize"
        role="button"
        style="width: 32px; height: 32px;"
        tabindex="0"
      />
      <div
        aria-label="close"
        class="apptentive-appbar__action apptentive-appbar__action--close"
        id="message-center__action--close"
        role="button"
        style="width: 32px; height: 32px;"
        tabindex="0"
      />
    </div>
  </div>
  <div
    class="apptentive-interaction__confirm-wrapper opened"
  >
    <span>
      Progress will be lost. Would you like to exit?
    </span>
    <button
      class="apptentive-interaction__confirm-wrapper__ok"
    >
      OK
    </button>
    <button
      class="apptentive-interaction__confirm-wrapper__cancel"
    >
      CANCEL
    </button>
  </div>
  <div
    class="apptentive-message-center__content"
  >
    <div
      class="apptentive-intro"
    >
      <p
        class="apptentive-typography__lead"
      >
        Please let us know how to make Apptentive better for you!
      </p>
    </div>
    <div
      class="apptentive-message-center-greeting"
    >
      <h2
        class="apptentive-message-center-greeting__title apptentive-typography__header"
      >
        Hello!
      </h2>
      <p
        class="apptentive-typography__body"
      >
        We'd love to get feedback from you on our app. The more details you can provide, the better.
      </p>
    </div>
    <div
      class="apptentive-message-center-feedback"
    >
      <textarea
        class="apptentive-textinput"
        name="response"
        placeholder="Please leave detailed feedback"
        required=""
        rows="4"
      />
    </div>
    <div
      class="apptentive-message-center-profile"
    >
      <h2
        class="apptentive-message-center-profile__title apptentive-typography__body"
      >
        Who are we speaking with?
      </h2>
      <label
        class="apptentive-message-center-profile__name"
      >
        <input
          class="apptentive-textinput"
          placeholder="Name"
          type="text"
        />
      </label>
      <label
        class="apptentive-message-center-profile__email"
      >
        <input
          class="apptentive-textinput"
          placeholder="Email"
          required=""
          type="email"
        />
      </label>
    </div>
    <div
      class="apptentive-message-center__actions"
    >
      <button
        class="apptentive-button--primary"
        type="button"
      >
        Send
      </button>
      <apptentive-errors
        class="apptentive-errors"
        style="display: none;"
      />
    </div>
    <div
      class="apptentive-interaction__overlay"
    />
  </div>
  <apptentive-message-center-thank-you
    class="apptentive-message"
    style="display: none;"
  >
    <div
      class="apptentive-message__content-wrapper"
    >
      <p
        class="apptentive-message__content"
      >
        We will get back to you soon!
      </p>
    </div>
    <div
      class="apptentive-message__actions"
    >
      <button
        class="apptentive-button--primary"
        title="Close"
      >
        Close
      </button>
    </div>
  </apptentive-message-center-thank-you>
  <footer
    class="apptentive-footer"
  />
</apptentive-message-center>
`;
